README.md

```diff
@@ -1,18 +1,74 @@
+
+This is the ~~first~~ second attempt to create a simple wrapper around the declarative
 parts of kubectl.
 
-## In a nut-shell
+## The Goal
+
+We want to consider our declarative configuration in Nix (committed to git) as, 
+the single (and only) source of truth.
+
+This means that whenever we ask for state in our kubernetes clusters to be applied,
+any cluster object in any namespace that is managed by kubernixos should:
+
+1) (if not in cluster) be created.
+
+2) (if needed) be updated to match the state of our Nix closure.
+
+3) (if in cluster, but not in Nix closure) be terminated.
+
+`kubectl apply` takes good care of bullet points 1 and 2, but `kubectl apply --prune` 
+is unfortunately not mature enough to serve our needs. Thus;
+the prune part of the apply command in particular has been re-implemented in this project
+using the kube client apis for Golang.  
+
+## How it works
+
+Kubernetes objects are **not** managed by kubernixos, 
+unless they are labelled with the "kubernixos" label.
+
+The value of this label contains a sha256 checksum of the entire deployment JSON-blob.
+The purpose of this is to be able to identify objects for pruning.
+If the checksum of the current deployment does not match the object checksum, it is concidered "unwanted"; ready for pruning. 
+
+These are the stages involved in a kubernixos rollout to a cluster:
+
+1) Invoke the NixOS module system through `nix eval` to get the value of the module
+attribute `config.kubernixos`.
 
-Basically the "kubernixos" script is nothing more than:
-`echo $BIG_JSON_BLOB | kubectl apply -f - -l reconciler=kubernixos --prune`
+2) Create a kubernetes auth config using the local machine kubeconfig.
+(~/.kube/config or KUBECONFIG in the environment)
 
-Note the label selector and `--prune` in particular.
-This is what ensures that all objects in the cluster, which are **not** present
-in the JSON blob get deleted (pruned). This is what makes kubernixos declarative.
-In that; only what is declared in the configuration is allowed to
-exist in the cluster.
+3) Query the total list of available API-resources on the Kubernetes API-server.
+
+4) Apply the value of `config.kubernixos.manifests` through `kubectl apply`.
+
+5) Get all kubernetes objects in all namespaces of all types, having the label "kubernixos".
+
+6) Compare the "kubernixos"-label value with that of `config.kubernixos.checksum`,
+and delete (prune) objects that has the label, but with a different value. 
+
+## Usage
+
+`kubernixos`: Everything is eval'ed and checked, but nothing is outputted nor mutated.
+
+`kubernixos dump`: Will dump the JSON-blob generated by kubernixos, ready for kubectl apply.
+This will mutate nothing.
+
+`kubernixos apply`: Like kubectl apply. This will apply kubernixos state to the cluster.
+Unwanted objects will not be pruned, unless "prune" is added to the command explicitly.
```

Comment:
Is `kubernixos apply --prune` a thing?

Reply:
Yes.. But here be dragons. If you do that, you opt in to the prune mechanism of kubectl, which.. you probably don't want :P Do you think we should block this somehow?